File: src/inPython/inheritance/Inheritance2.py
# package python.inheritance

# Eliminating redundant code using a common superclass
# 
# To be continued: See Inheritance3
def inheritance_program():
    _pet = Pet("any", 0)  # Ok, but senseless
    dog = Dog("Lassie", 4)
    cat = Cat("Misse", 5, False)

    # Use inherited methods
    print(f"{dog.get_name()}:{dog.get_age()}")
    print(f"{cat.get_name()}:{cat.get_age()}")


class Pet:
    def __init__(self, name: str, age: int):
        self.name = name  # Everything common for Cat and Dog here
        self.age = age

    def get_name(self):
        return self.name

    def get_age(self):
        return self.age


# Dog is subclass to Pet, will inherit all from super type
class Dog(Pet):
    def __init__(self, name: str, age: int):
        super().__init__(name, age)

    # get_name() and get_age() inherited but invisible


class Cat(Pet):
    def __init__(self, name: str, age: int, is_evil: bool):
        super().__init__(name, age)
        self.is_evil = is_evil

    # get_name() and get_age() inherited but invisible

    def is_evil(self):
        return self.is_evil


if __name__ == "__main__":
    inheritance_program()

File: src/inJava/inheritance/Inheritance2.java
package inJava.inheritance;

import static java.lang.System.out;

// Eliminating redundant code using a common superclass
//
// To be continued: See Inheritance3
class Inheritance2 {
    static void inheritanceProgram() {
        Pet _pet = new Pet("any", 0); // Passes type checker, but senseless
        Dog dog = new Dog("Lassie",4);
        Cat cat = new Cat("Misse",5,false);

        // Use duplicated methods
        out.println(dog.getName() + ":" + dog.getAge());
        out.println(cat.getName() + ":" + cat.getAge());
    }

    static class Pet {
        private String name;
        private int age;

        Pet(String name, int age){
            this.name = name;
            this.age = age;
        }

        String getName() {
return name;
}

        int getAge() {
return age;
}
    }

    static class Dog extends Pet {
        Dog(String name, int age) {
            super(name, age);
        }

        // getName() and getAge() inherited but invisible
    }

    static class Cat extends Pet {
        private boolean isEvil;

        Cat(String name, int age, boolean isEvil) {
            super(name, age);
            this.isEvil = isEvil;
        }

        // getName() and getAge() inherited but invisible

        boolean isEvil() {
            return this.isEvil;
        }

    }
    public static void main(String[] args) {
        inheritanceProgram();
    }
}
